src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PENamedTypeSymbol.cs

```diff
@@ -2621,14 +2622,14 @@ private void EnsureTypeParametersAreLoaded()
                     // If this is a nested type generic parameters in metadata include generic parameters of the outer types.
                     int firstIndex = _genericParameterHandles.Count - _arity;
 
-                    TypeParameterSymbol[] ownedParams = new TypeParameterSymbol[_arity];
-                    for (int i = 0; i < ownedParams.Length; i++)
+                    using var ownedParams = TemporaryArray<TypeParameterSymbol>.GetInstance(_arity);
```

Comment:
This is not based on anything I seen in a trace, but I think it makes sense.

Reply:
note: this is a base better served by using an ArrayBuilder with an known count, and `ToImmutableAndFree()`.  It will always make the right-sized array, and alway pass that into the ImmutableArray withotu a copy.  TempArray will still have a copy in the 'small number of elements' case.  And it is better suited when either the count isn't known, or hte array will often be thrown away and we want to keep things on the stack.

Reply:
> My intent is _arity in most cases will be <= 4, so that keeps everything on the stack, and only have a single allocation for the immutable array

ArrayBuilder is still good there.  You'll provide the exact length, so even in the 1, 2, 3 cases it will just allocate exactly that array (which you *always need no matter what*), fill it, and then *move* that array into the ImmutableArray result.

Reply:
> Thanks!! I'm curious to know the alternatives!

The ideal solution if `_genericParameterHandles` was an immutable array would have been [`CreateRange<TSource,TResult>(ImmutableArray<TSource>, Int32, Int32, Func<TSource,TResult>)`](https://learn.microsoft.com/en-us/dotnet/api/system.collections.immutable.immutablearray.createrange?view=net-7.0#system-collections-immutable-immutablearray-createrange-2(system-collections-immutable-immutablearray((-0))-system-int32-system-int32-system-func((-0-1)))).

I think I'd use this implementation, and update the caller to return the result directly:

```cs
private ImmutableArray<TypeParameterSymbol> EnsureTypeParametersAreLoaded()
{
    return InterlockedOperations.InterlockedInitialize(
        ref _lazyTypeParameters,
        static self =>
        {
            if (self._arity == 0)
                return ImmutableArray<TypeParameterSymbol>.Empty;

            var moduleSymbol = self.ContainingPEModule;

            // If this is a nested type generic parameters in metadata include generic parameters of the outer types.
            int firstIndex = self._genericParameterHandles.Count - self._arity;

            var ownedParams = ArrayBuilder<TypeParameterSymbol>.GetInstance();
            ownedParams.Count = self._arity;
            for (int i = 0; i < ownedParams.Count; i++)
            {
                ownedParams[i] = new PETypeParameterSymbol(moduleSymbol, self, (ushort)i, self._genericParameterHandles[firstIndex + i]);
            }

            return ownedParams.ToImmutableAndFree();
        },
        this);
}
```